### PR TITLE
ICollection -> folderish

### PR DESCRIPTION
**Pull request by kiorky**
cc @ebrehault @pbauer

### CODE REVIEW

**Comment by kiorky** [comment]
tied to plone/plone.app.content/pull/25

File: plone/app/layout/globals/context.py [hunk 1]
```diff
@@ -163,7 +169,14 @@ def folder(self):
...
     def is_folderish(self):
-        return bool(getattr(aq_base(aq_inner(self.context)), 'isPrincipiaFolderish', False))
+        cond1 = bool(
+            getattr(aq_base(aq_inner(self.context)),
+                    'isPrincipiaFolderish', False))
+        cond2 = False
+        if HAS_PAC:
+            if ICollection.providedBy(self.context):
+                cond2 = True
+        return cond1 or cond2
```
Comment (thet) [inline plone/app/layout/globals/context.py]: collection are not folderish per default, AFAIK (see: https://github.com/plone/plone.app.contenttypes/blob/master/plone/app/contenttypes/content.py - it derives from ``plone.dexterity.content.Item``). @pbauer - your input ?

Comment (kiorky) [inline plone/app/layout/globals/context.py]: Yep, but we can call folderContents & co on it.
And we must keep that behavior as many addons, like eea.facetednaviguation relies on it
cc @avoinea 

This bit of hack is i think just sufficient to support new Collections.

Comment (pbauer) [inline plone/app/layout/globals/context.py]: @kiorky Can you explain how eea and other addons rely on collections being folderish even though they are in fact not? I don't think this hack is a great solution, instead eea should use 'results' if it uses new collections to access the content.

Comment (tisto) [inline plone/app/layout/globals/context.py]: @kiorky This doesn't look right to me. I think you should try to solve this issue on another level (or give us a better explanation why this is really needed).

Comment (kiorky) [inline plone/app/layout/globals/context.py]: Collections were always somewhat folderish... i dont understand you...
It s a great loss of a feature if we loose that ...

Comment (tisto) [inline plone/app/layout/globals/context.py]: No. Folderish collections have been removed in Plone 4.0/4.2: http://plone.293351.n2.nabble.com/Nested-collection-removal-in-Plone-4-0-4-2-td7556157.html

Comment (kiorky) [inline plone/app/layout/globals/context.py]: I just want getFolderContents, if not plone is  broken.

Comment (kiorky) [inline plone/app/layout/globals/context.py]: To try to reexplain:

With the current plone.app.contenttypes, you can't go on the "contents" tab of a collection.

We were used for a lot of people that the "contents" tab of a collections just list the items that the collection's query bring out.

Thus, you can easily do some managment views, using portal_actions to add a button to push those contents via a post on those. This is easy, and fast, and this is there for years.

The idea is just to have the "contents" tab available, which behaves like today, even if we can't physically add content to a collection, the collection is an hybrid container in the sense that its content is not physically "inside", but it has definitivly a folderish behavior in the sense that we should be able to do actions on the content that it has selected via its criterions.

Problem now, is that on p.a.contenttypes collection, that's not the case without my patches or something similar.

**Comment by kiorky** [comment]
(and that's why i didnt made it implement any folderish interface, to let it not be a "real container")

File: plone/app/layout/globals/context.py (hunk 1 above)
Comment (kiorky) [inline plone/app/layout/globals/context.py]: (and that's why i didnt made it implement any folderish interface, to let it not be a "real container")

Comment (pbauer) [inline plone/app/layout/globals/context.py]: Does having a contents-tab make any sense for collections? What can you do on it?

Comment (kiorky) [inline plone/app/layout/globals/context.py]: did you really read what i wrote,  @pbauer ? ;)

**Comment by thet** [comment]
i think your proposal abuses the collection content type. collections are not folderish by intention, and if they were (actually you can - for example - add the collection behavior to a folder), the contents tab shouldn't display the collection results but what's physically contained.

i suggest, you better implement a special results view based on the foldercontents view or mockup's filemanager pattern.

so, i'm -1 for this and related pull requests.

**Comment by kiorky** [comment]
@thet not at all, the folderish behavior is NOT at all what i want as it:
* will let users add content in this collection
* list content on the collection

The old AND correct behavior is to not add content but list content selected via the criterias, comeon, this is there from ages.....

**Comment by thet** [comment]
i'm not sure, if the old behavior it's the "correct" behavior.
in my opinion, the folder contents should list what's physically contained. listing the collection criteria's result is confusing, in my opinion, and adds complexity.

i understand the usefulness of your proposal, but there should be another way to implement it. what about an explicit ``@@resultedit`` view, which just uses the [``structure``](http://plone.github.io/mockup/dev/#pattern/structure) pattern from mockup (not the ``filemanager`` as i mentioned before)?

**Comment by kiorky** [comment]
The problem is more in code portability and compatibility across versions, even major.
portal_actions+folder_contents and alike were a quick way to make mass "actions", and moreover, "custom actions" on folderish and alike contents, like they just brand :). 
Even with a structure integration, it will _break_ custom code, and badly where we could make it just work.

You have a container, items are listed, either from a real physical POV in case of any folderish which is not a Collection, and for Collection, you get the content from the collection seems just reasonable and we should really be really more careful about changing such a behavior which is there now from likely a decade. I'm not saying that a first it was the most usable, but now it's there and our users, including us are used to it, and (custom*) code has been designed around it.

I do not see the point of breaking features, even more intentionally.

File: plone/app/layout/globals/context.py (hunk 1 above)
Comment (pbauer) [inline plone/app/layout/globals/context.py]: Ups, sorry, I totally missed that :-) I'll have a look at folder_contents in collections in the next days.

**Comment by tisto** [comment]
@kiorky You can not expect full code portability between major Plone versions. Furthermore I don't think it is too hard to add the folder listing tab to collections in your custom code. I'm still -1.

**Comment by kiorky** [comment]
FolderContents views cannot work without is_folderish returning True.

**Comment by tisto** [comment]
Can't you just override is_folderish or create an entirely separated view?

**Comment by kiorky** [comment]
Did you read the whole foldecontents stack, it is not really easy

**Comment by tisto** [comment]
That might be true and if we can make things easier there, we are open for suggestions. Though, you have to understand that we can not introduce unneccessary complexity into Plone core code, just to support a use case that maybe one or two percent of all integrators have. If we would do so we would end up with unmaintainable code pretty soon.

If you look at the link to the mailinglist I posted earlier, you will see that I had a similar use case for folderish collections, but in the end solved it in my custom code because my use case differed from what most Plone integrators need.

**Comment by djowett** [comment]
Hi, as a humble integrater I just want to say thanks to @tisto for that approach - though I'm sure it's painful for some in situations like this & his earlier one, the less wierd or unpredictable stuff happening in the core, the better.

**Comment by kiorky** [comment]
this isnt unpredictable for years  and the code which is there reacts specifically, there isnt floor for unpredicted behavior ;).

On a call with @ebrehault (IRL) the only case to make room for discussions is how to deal with the included pull requests and the activated folderish behavior toward the folder_contents items to display.

Breaking an available feature which can be taken care of with a few line of proxy code should not make room for such a discussion, more over if we have spotten it because it reach release state. Which is the case here.

**Comment by davisagli** [comment]
I disagree -- with that approach Plone can only ever become more complex. If there's a feature that only a few people use and it's easy to turn it into an add-on, that's a better approach.

Also note that the new folder contents in Plone 5 now includes the ability to query. I realize it can't save that query like a collection does, but again: good add-on candidate.

**Comment by kiorky** [comment]
Why more complex, it will just be identical from the user POV ;).

Even for users im talking about, which were used to manage their content via well defined collections + actions, they will have to redefine all their workflows etasblished for years.

For me, it will be just a deal with the client to migrate this view, and we have here the in depth technical overview to make this change.
However, this may not be the case of any integrator that hacked something with portal_actions, and maybe even not views, but python scripts or other oldities.

Indeed, For my specific usecase, yes I think i'm just frustrated on this but can move on to do a foldercontent alike view that list items in the folder/aggregated via criteria in case of collections that post to the other existing view, i still have to sell it but yes it can be done and that another question anyway.

**Comment by jean** [comment]
Hi all

I agree of course that we don't want a crufty Plone with lots of exceptions.

But it feels quite wrong to me to have a collection which is not folderish.
Being able to drop in a collection (for reading) anywhere, in place of a
folder, is a powerful concept.

Having to remember that you need to get items from a folder in one way,
and from a collection in another way, moves the ugliness from the code
to the conceptual level.

Best of course would be a solution that satisfies both those looking from
the code point of view and the integrator point of view.

-- 
jean                                              . .. .... //\\\oo///\\

**Comment by tisto** [comment]
@jean Folderish collections have been removed because there are serious UI problems with it. See the link I posted above for details. Collections are already quite complex for end users. Adding more functionality and complexity just won't work in my opinion. Do you really want to expose a nested persistent search to the end user that inherit its parents search criteria? Do you think we can come up with a sane UI for that? Do you think you can explain those concepts to your end users?

If we don't learn from the past, we are going to repeat our mistakes over and over again. We already had folderish collections, it did not work. If we now re-add them, I'm sure we want to get rid of them in future again.

**Comment by ebrehault** [comment]
Something to have in mind: with Plone 5, we will offer the possibility to turn *any* content type into folderish.
So basically, if you turn your collection into folderish, it means you can add (physically) some contents into it, which is very nice (typical use case: I want a News folder, that would contain the news, and that will show those news with a given sorting order and a given filtering, instead of the usual Folder + Collection + Make it default page for folder, I just create a collection and put my news item in it, that's much better).
In order to run this use case properly, we do need the Folder content page to work as expected (=show the physically contained items), so it makes the old behaviour (=show the collection virtual items) not applicable anymore.
If this old behavior is still needed in some cases, I think an add-on might provide it very easily as @davisagli said.

**Comment by tisto** [comment]
@ebrehault Are you talking about future Plone 5.x versions? We just removed the folderish profile from p.a.contenttypes and postponed the folderish PLIP to Plone 5.1, right? You can, of course, turn every content type into a folderish type already by just replacing its base class or using collective.folderishtypes.

**Comment by thet** [comment]
@ebrehault @tisto in plone.app.contenttypes we now have a collection behavior, which can give anything collection-ish functionality - also folders.

@ebrehault you're saying that ``folder_contents`` should show, what's physically contained (i totally agree). but isn't the purpose of this and related pull-requests to return the collection results on ``folder_contents`` (which is confusing and wrong in my opinion)?

**Comment by jensens** [comment]
Unfortunately this one seems a bit controversal. Also there the folderish types plip. I close this one, if you think this is wrong please reopen and comment.

**Pull request closed without merging**